File: bench/diff_glob.c
// SPDX-License-Identifier: (GPL-2.0 OR MIT)
// Tier-2.5 differential oracle: C original vs Rust translation, glob.
// Reference extracted from lib/glob.c (v7.1), kept byte-identical.
//
// The real upstream KUnit suite (lib/tests/glob_kunit.c, 64 cases) is
// the primary correctness gate for glob_match's tricky backtracking
// logic and already passes on real boot (dev.py check). This oracle's
// job is narrower: fuzz glob_match_len specifically -- the
// length-bounded variant KUnit doesn't cover at all -- exercising
// truncation mid-pattern and embedded NUL bytes glob_match's plain
// NUL-terminated contract can't reach.
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static bool glob_match_str(char const *pat, char const *str, char const *str_end)
{
	char const *back_pat = NULL, *back_str = NULL;

	for (;;) {
		unsigned char c = (str_end && str >= str_end) ? '\0' : *str;
		unsigned char d = *pat++;

		str++;

		switch (d) {
		case '?':
			if (c == '\0')
				return false;
			break;
		case '*':
			if (*pat == '\0')
				return true;
			back_pat = pat;
			back_str = --str;
			break;
		case '[': {
			if (c == '\0')
				return false;
			bool match = false, inverted = (*pat == '!');
			char const *class = inverted ? pat + 1 : pat;
			unsigned char a = *class++;

			do {
				unsigned char b = a;

				if (a == '\0')
					goto literal;

				if (class[0] == '-' && class[1] != ']') {
					b = class[1];

					if (b == '\0')
						goto literal;

					class += 2;
				}
				if (a <= c && c <= b)
					match = true;
			} while ((a = *class++) != ']');

			if (match == inverted)
				goto backtrack;
			pat = class;
			}
			break;
		case '\\':
			d = *pat++;
			// fallthrough
		default:
literal:
			if (c == d) {
				if (d == '\0')
					return true;
				break;
			}
backtrack:
			if (c == '\0' || !back_pat)
				return false;
			pat = back_pat;
			str = ++back_str;
			break;
		}
	}
}

static bool glob_match_len(char const *pat, char const *str, size_t len)
{
	return glob_match_str(pat, str, str + len);
}

// Explicit LCG (same constants used across all bench/diff_*.c files).
static uint64_t lcg_state;
static uint32_t lcg_next(void)
{
	lcg_state = lcg_state * 6364136223846793005ULL + 1442695040888963407ULL;
	return (uint32_t)(lcg_state >> 32);
}

// Pattern/string alphabets kept small and metacharacter-heavy so most
// random draws actually exercise ?, *, [ ] backtracking rather than
// degenerating into all-literal comparisons.
static const char alphabet[] = "ab[]!*?\\-c";

static void gen_buf(char *buf, size_t max_len, size_t *out_len)
{
	size_t n = lcg_next() % (max_len + 1);
	for (size_t i = 0; i < n; i++)
		buf[i] = alphabet[lcg_next() % (sizeof(alphabet) - 1)];
	// Occasionally embed a real NUL mid-buffer to exercise str_end's
	// "NUL within len still terminates" contract explicitly.
	if (n > 2 && lcg_next() % 4 == 0)
		buf[n / 2] = '\0';
	*out_len = n;
}

int main(int argc, char **argv)
{
	long n = argc > 1 ? atol(argv[1]) : 5000;
	lcg_state = argc > 2 ? (uint64_t)atol(argv[2]) : 271828;
	char pat[17];
	char str[17];

	for (long i = 0; i < n; i++) {
		size_t pat_len, str_len;
		gen_buf(pat, 16, &pat_len);
		pat[pat_len] = '\0'; // pat is always the real NUL-terminated contract
		gen_buf(str, 16, &str_len);

		// len is sometimes the real buffer length, sometimes
		// deliberately truncated shorter, to exercise mid-pattern
		// cutoff.
		size_t len = (lcg_next() % 2 == 0) ? str_len
						    : (lcg_next() % (str_len + 1));

		bool r = glob_match_len(pat, str, len);
		printf("glob_match_len,%zu,%zu,%d\n", pat_len, len, r);
	}
	return 0;
}

File: bench/diff_glob.rs
// SPDX-License-Identifier: (GPL-2.0 OR MIT)
//! Tier-2.5 differential oracle: Rust translation side, glob. Faithful
//! copy of lib/glob_rs.rs's glob_match_str/glob_match_len algorithm
//! (kernel-crate bindings/no_mangle stripped) — same protocol/LCG as
//! diff_glob.c. See diff_glob.c's header comment for why this oracle
//! targets glob_match_len specifically (the real upstream KUnit suite,
//! lib/tests/glob_kunit.c, already covers glob_match's 64 cases on
//! real boot).

unsafe fn glob_match_str(pat: *const u8, str_: *const u8, str_end: Option<*const u8>) -> bool {
    let mut back_pat: Option<*const u8> = None;
    let mut back_str: Option<*const u8> = None;

    let mut pat = pat;
    let mut str_ = str_;

    loop {
        let at_end = str_end.is_some_and(|end| str_ >= end);
        let c: u8 = if at_end { 0 } else { unsafe { *str_ } };
        let d: u8 = unsafe { *pat };
        pat = unsafe { pat.add(1) };
        str_ = unsafe { str_.add(1) };

        match d {
            b'?' => {
                if c == 0 {
                    return false;
                }
            }
            b'*' => {
                if unsafe { *pat } == 0 {
                    return true;
                }
                back_pat = Some(pat);
                str_ = unsafe { str_.sub(1) };
                back_str = Some(str_);
            }
            b'[' => {
                if c == 0 {
                    return false;
                }
                let mut is_match = false;
                let inverted = unsafe { *pat } == b'!';
                let mut class = if inverted { unsafe { pat.add(1) } } else { pat };
                let mut a: u8 = unsafe { *class };
                class = unsafe { class.add(1) };

                let mut malformed = false;
                loop {
                    let mut b = a;

                    if a == 0 {
                        malformed = true;
                        break;
                    }

                    let class0 = unsafe { *class };
                    if class0 == b'-' {
                        let class1 = unsafe { *class.add(1) };
                        if class1 != b']' {
                            b = class1;
                            if b == 0 {
                                malformed = true;
                                break;
                            }
                            class = unsafe { class.add(2) };
                        }
                    }
                    if a <= c && c <= b {
                        is_match = true;
                    }

                    a = unsafe { *class };
                    class = unsafe { class.add(1) };
                    if a == b']' {
                        break;
                    }
                }

                if malformed {
                    if c == d {
                        if d == 0 {
                            return true;
                        }
                        continue;
                    }
                    if c == 0 || back_pat.is_none() {
                        return false;
                    }
                    pat = back_pat.unwrap();
                    str_ = unsafe { back_str.unwrap().add(1) };
                    back_str = Some(str_);
                    continue;
                }

                if is_match == inverted {
                    if c == 0 || back_pat.is_none() {
                        return false;
                    }
                    pat = back_pat.unwrap();
                    str_ = unsafe { back_str.unwrap().add(1) };
                    back_str = Some(str_);
                    continue;
                }
                pat = class;
            }
            b'\\' => {
                let d = unsafe { *pat };
                pat = unsafe { pat.add(1) };
                if c == d {
                    if d == 0 {
                        return true;
                    }
                    continue;
                }
                if c == 0 || back_pat.is_none() {
                    return false;
                }
                pat = back_pat.unwrap();
                str_ = unsafe { back_str.unwrap().add(1) };
                back_str = Some(str_);
            }
            _ => {
                if c == d {
                    if d == 0 {
                        return true;
                    }
                    continue;
                }
                if c == 0 || back_pat.is_none() {
                    return false;
                }
                pat = back_pat.unwrap();
                str_ = unsafe { back_str.unwrap().add(1) };
                back_str = Some(str_);
            }
        }
    }
}

fn glob_match_len(pat: &[u8], str_: &[u8], len: usize) -> bool {
    unsafe {
        let str_ptr = str_.as_ptr();
        glob_match_str(pat.as_ptr(), str_ptr, Some(str_ptr.add(len)))
    }
}

// Identical LCG to diff_glob.c.
struct Lcg(u64);
impl Lcg {
    fn next(&mut self) -> u32 {
        self.0 = self
            .0
            .wrapping_mul(6364136223846793005)
            .wrapping_add(1442695040888963407);
        (self.0 >> 32) as u32
    }
}

const ALPHABET: &[u8] = b"ab[]!*?\\-c";

fn gen_buf(rng: &mut Lcg, max_len: usize, buf: &mut [u8; 17]) -> usize {
    let n = (rng.next() as usize) % (max_len + 1);
    for i in 0..n {
        buf[i] = ALPHABET[(rng.next() as usize) % ALPHABET.len()];
    }
    if n > 2 && rng.next() % 4 == 0 {
        buf[n / 2] = 0;
    }
    n
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let n: u64 = args.get(1).and_then(|a| a.parse().ok()).unwrap_or(5000);
    let seed: u64 = args.get(2).and_then(|a| a.parse().ok()).unwrap_or(271828);
    let mut rng = Lcg(seed);

    for _ in 0..n {
        let mut pat_buf = [0u8; 17];
        let pat_len = gen_buf(&mut rng, 16, &mut pat_buf);
        pat_buf[pat_len] = 0; // pat is always the real NUL-terminated contract

        let mut str_buf = [0u8; 17];
        let str_len = gen_buf(&mut rng, 16, &mut str_buf);

        let len = if rng.next() % 2 == 0 {
            str_len
        } else {
            (rng.next() as usize) % (str_len + 1)
        };

        let r = glob_match_len(&pat_buf[..=pat_len], &str_buf, len);
        println!("glob_match_len,{},{},{}", pat_len, len, r as i32);
    }
}
